README.md
bkp_cli
=======

GUI/Console based backup utility; works on both Linux and Windows. The script actually creates copies of selected
folders and files to a pre-defined destination (which is a folder named as backup location). Subsequent updates will
cause only the modified folders and files to be backed up, thereby reducing app load and processing time.


changelog.txt
Version 2.0

1. Improved algorithm to update backup without creating or reading configuration file.
2. Creation and updation of backup is now using only one command line parameter, '-a'
3. Completely revamped the GUI interface
4. Trying out git ...
5. Adding a new file



abt.py
#!/usr/bin/env python

'''
    GUI interface for managing backups
    Copyright (C) 2012  Pradeep Balan PIllai

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

'''
abt.py
Version: 2.01
Module to create About dialog box
'''

import wx
import wx.html


class About(wx.Frame):
    def __init__(self, parent=None):
        wx.Frame.__init__(self, parent, style=wx.DEFAULT_FRAME_STYLE &
                          ~ wx.MAXIMIZE_BOX)
        self.draw_gui()

    def draw_gui(self):
        sizer = wx.BoxSizer(wx.VERTICAL)
        self.SetSizer(sizer)
        book = wx.Notebook(self)        
        content = wx.html.HtmlWindow(book)
        book.AddPage(content, 'About')
        content.LoadPage('abt.html')
        license = wx.TextCtrl(book, style=wx.TE_MULTILINE|wx.TE_READONLY)
        book.AddPage(license, 'License')
        license.LoadFile('license.txt')

        ok_button = wx.Button(self, label='OK')
        ok_button.Bind(wx.EVT_BUTTON, self.click_ok_button)

        sizer.Add(book, proportion=1, flag=wx.EXPAND)
        sizer.Add(ok_button, flag=wx.ALIGN_CENTRE_HORIZONTAL)
        self.Show()

    def click_ok_button(self, event):
        self.Destroy()

if __name__ == '__main__':
    app = wx.App()
    About()
    app.MainLoop()




bkp.py
#!/usr/bin/env python

'''
    Initialisation script for command line based backup utility... V 1.0
    Copyright (C) 2012 Pradeep Balan Pillai

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

'''
bkp.py:
Command line interface for managing backups. Currently this has got
limited functionality
Version: 2.0

'''
import sys

import backup

# Function to process command
def process_command(command_list):
    try:
        file_handler = open(command_list[2], 'r')
        # Remove the newline character also while adding path list
        data = [line.rstrip() for line in file_handler.readlines()]
        #data = file_handler.readlines()
        file_handler.close()
    except IOError:
        print 'Error reading data file'

    # Remove the newline characters
    
    if command_list[1] == '-n':
        #backup.create_new_backup(data[0])
        pass
    elif command_list[1] == '-a':
        #backup.add_items_to_backup(data[1:], data[0])
        backup.update_backup_location(data[1:], data[0])
    elif command_list[1] == '-t':
        pass # test code here


if len(sys.argv) < 3:
    print 'Insufficient parameters'
else:
    process_command(sys.argv)


backup.py
#!/usr/bin/env python

'''
    Module defining functions for a simple backup utility..V 1.01
    Copyright (C) 2012  Pradeep Balan PIllai

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

'''
backup.py
Module for performing various backup related activities
Version: 2.0
'''

import os
import shutil
import time



# Function updates the file at backup location
def update_file(path, bkp_location):
    npath = os.path.normpath(path)
    nbkp_location = os.path.normpath(bkp_location)
    arch_path = os.path.join(nbkp_location, os.path.split(npath)[1])
    update_path(npath, arch_path)


# Function updates folder at backup location
def update_folder(path, bkp_location):
    npath = os.path.normpath(path)
    nbkp_location = os.path.normpath(bkp_location)
    
    tree = os.walk(npath)
    base = os.path.split(npath)[0]

    for item in tree:
        for name in item[2]:
            abs_path = os.path.normpath(os.path.join(item[0], name))
            arch_path = os.path.normpath(os.path.join(nbkp_location,
                                            abs_path.lstrip(base)))

            print 'abs:%s arch:%s' % (abs_path, arch_path)
            update_path(abs_path, arch_path)

# Function for updatibg the path
def update_path(abs_path, arch_path):
    nabs_path = os.path.normpath(abs_path)
    narch_path = os.path.normpath(arch_path)
    print 'Checking: %s' % nabs_path
    if os.path.exists(narch_path) and (os.path.getmtime(nabs_path)
                                    < os.path.getmtime(narch_path)):
        pass
    else:
        # Create sub-directories if required
        try:
            os.makedirs(os.path.split(narch_path)[0])
        except os.error:
            pass
        
        # Copy file path
        try:
            print 'Adding: %s...%d' % (nabs_path, os.path.getsize(nabs_path))
            shutil.copy(nabs_path, narch_path)
        except IOError:
            print 'Error while handling file'

# Function to add or update files/ folders to backup location
def update_backup_location(path_list, bkp_location):
    for path in path_list:
        if os.path.isfile(path):
            update_file(os.path.normpath(path), os.path.normpath(bkp_location))
        elif os.path.isdir(path):
            update_folder(os.path.normpath(path),
                          os.path.normpath(bkp_location))
    print '\n' + 10*'=' + ' DONE ' + 10*'='            

# Function to create an empty data file at backup location
def create_data_file(bkp_location):
    file_handle = open(os.path.join(bkp_location, 'bkpconfig.txt'), 'w')
    file_handle.write(os.path.normpath(bkp_location)+'\n')
    file_handle.close()

# Function to add single path to data file
def add_path_to_data_file(bkp_path, source_path):
    try:
        file_handle = open(os.path.join(os.path.normpath(bkp_path),
                           'bkpconfig.txt'),'a')
        file_handle.write(os.path.normpath(source_path) + '\n')
        file_handle.close()
    except IOError:
        print 'Error opening config data file; check backup path'


# Function to add multiple paths to data file at a time
def add_paths_to_data_file(bkp_path, source_paths):

    try:
        file_handle = open(os.path.join(os.path.normpath(bkp_path),
                           'bkpconfig.txt'),'a')
        for path in source_paths:
            file_handle.write(os.path.normpath(path) + '\n')
        file_handle.close()
    except IOError:
        print 'Error opening config data file; check backup path'
  
# Function to update backup when only backup location is passed as an argument
def update_all(bkp_location):
    try:
        file_handler = open(os.path.join(os.path.normpath(bkp_location),
                            'bkpconfig.txt'), 'r')
        path_data = [line.rstrip() for line in file_handler.readlines()]
        file_handler.close()
        update_backup_location(path_data[1:], bkp_location)
    except IOError:
        print 'Error opening config data file; check backup path.'

'''
Functions returns True if the passed path is a valid backup location
by examining the presence of config file
'''
def is_valid_backup(path):
    pass
    config_file_path = os.path.join(os.path.normpath(path), 'bkpconfig.txt')
    if os.path.exists(config_file_path):
        return True
    else:
        return False


''' 
Function to return the file path contained in the config file
'''
def get_backup_sources(bkp_location):
    config_file = os.path.join(os.path.normpath(bkp_location), 'bkpconfig.txt')
    try:
        file_h = open(config_file, 'r')
        source_paths = [line.rstrip() for line in file_h.readlines()]
        file_h.close()
        return source_paths[1:]

    except:
        print 'Configuration File Error'

'''
Function to update config file with new source data; return True if
update was successfull
'''
def update_config_file(bkp_location, source_paths):
    config_file = os.path.join(os.path.normpath(bkp_location), 'bkpconfig.txt')
    try:
        file_h = open(config_file, 'w')
        file_h.write(os.path.normpath(bkp_location) + '\n')
        for path in source_paths:
            file_h.write(os.path.normpath(path)+'\n')
        file_h.close()
        return True
    except:
        return False


itemdialog.py
#!/usr/bin/env python

'''
    GUI interface for managing backups
    Copyright (C) 2012  Pradeep Balan PIllai

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

'''
bkpgui.py
Version: 2.0
Module to create dialog box for visually modifying contents of a list
'''

import wx
import os

class ItemDialog(wx.Dialog):
    def __init__(self, parent=None, title=None, initial_list=[]):
        wx.Dialog.__init__(self, parent)
        self.item_list = initial_list
        self.draw_gui(title)
        #self.ShowModal()


    # Method created the widgets for main window
    def draw_gui(self, title):
        self.SetTitle(title)
        self.SetSize((400,300))


        # Main layout manger
        main_sizer = wx.BoxSizer(wx.VERTICAL)
        self.SetSizer(main_sizer)
        
        # Sub layout managers
        edit_sizer = wx.BoxSizer(wx.HORIZONTAL)
        edit_button_sizer = wx.BoxSizer(wx.VERTICAL)
        control_sizer = wx.BoxSizer(wx.HORIZONTAL)

        # main widgets
        command_text = wx.StaticText(self, label='Add files or folders:')
        self.item_display = wx.ListBox(self, style=wx.LB_MULTIPLE,
                                       choices=self.item_list)
        add_button = wx.Button(self, label='Add file')
        add_folder_button = wx.Button(self, label='Add folder')
        delete_button = wx.Button(self, label='Remove')
        delete_button.Enable(True)


        # Add respective widgets to sizers
        main_sizer.Add(command_text, flag=wx.EXPAND|wx.ALL, border=5)
        edit_sizer.Add(self.item_display, flag=wx.EXPAND, proportion=1,
                       border=5)
        edit_button_sizer.Add(add_button, border=5, flag=wx.ALL)
        edit_button_sizer.Add(add_folder_button, border=5, flag=wx.ALL)
        edit_button_sizer.Add(delete_button, border=5, flag=wx.ALL)
        edit_sizer.Add(edit_button_sizer, border=5, flag=wx.ALL)
        main_sizer.Add(edit_sizer, flag=wx.EXPAND|wx.ALL, 
                       proportion=1, border=5)

        control_sizer = self.CreateButtonSizer(wx.OK|wx.CANCEL)
        main_sizer.Add(control_sizer, flag=wx.EXPAND|wx.ALL, border=10)

        # Resize frame
        #self.Fit()
        self.Bind(wx.EVT_CLOSE, self.on_close, self)
        add_button.Bind(wx.EVT_BUTTON, self.click_add_file)
        add_folder_button.Bind(wx.EVT_BUTTON, self.click_add_folder)
        delete_button.Bind(wx.EVT_BUTTON, self.click_delete)

    # Event handler for OK button
    def click_ok(self, event):
        list = self.get_list()
        print list
        self.Destroy()


    # Manage close event of the dialog
    def on_close(self, event):
        self.Destroy()


    # Event handler for 'add file' button
    def click_add_file(self, event):
        dialog = wx.FileDialog(self, style=wx.FD_OPEN|wx.FD_MULTIPLE)
        resp = dialog.ShowModal()
        if resp == wx.ID_OK:
            selected_paths = dialog.GetPaths()
            self.add_items_to_list(selected_paths)
            dialog.Destroy()
        else:
            dialog.Destroy()

    # Event hanlder for 'add folder' button
    def click_add_folder(self, event):
        dialog = wx.DirDialog(self, style=wx.DD_DIR_MUST_EXIST)
        resp = dialog.ShowModal()
        if resp == wx.ID_OK:
            self.add_items_to_list([dialog.GetPath()])
            dialog.Destroy()
        else:
            dialog.Destroy()

    # Event handler for 'delete' button
    def click_delete(self, event):
        rem_index = self.item_display.GetSelections()
        print rem_index
        mod_list = self.delete_items_from_list(rem_index) 
        print mod_list
        self.item_display.Clear()
        self.item_display.InsertItems(mod_list, 0)

    # All items to list and update list display
    def add_items_to_list(self, items):
        for path in items:
            npath = os.path.normpath(path)
            self.item_list.append(str(npath))
            self.item_display.InsertItems([npath], len(self.item_list)-1) 

    # Return the updated list
    def get_list(self):
        return self.item_list


    # This method removes an existing item from the list and returns
    # the updated list. The indices of the item needs to be passed 
    # as an argument
    def delete_items_from_list(self, item_index):
        display_list = self.get_list()
        for i in item_index:
            display_list.pop(i)
        return display_list
    

if __name__ == '__main__':
    app = wx.App(redirect=False)
    ItemDialog(title='test dialog')
    app.MainLoop()



bkpgui.py
#!/usr/bin/env python

'''
    GUI interface for managing backups
    Copyright (C) 2012  Pradeep Balan PIllai

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
'''

'''
bkpgui.py
Version: 2.01
GUI front end for managing backups
'''


import wx

import backup
import itemdialog
import abt

class MainFrame(wx.Frame):
    def __init__(self, parent=None):
        wx.Frame.__init__(self, parent)
        self.draw_gui()
        self.Show()

    def draw_gui(self):
        self.SetTitle('Backup Utility')
        self.SetSize((500, 300))

        main_panel = wx.Panel(self)
        main_sizer = wx.BoxSizer(wx.VERTICAL)
        main_panel.SetSizer(main_sizer)

        select_label = wx.StaticText(main_panel, label='Select:')
        location_panel = wx.Panel(main_panel)
        edit_panel = wx.Panel(main_panel)
        control_panel = wx.Panel(main_panel)

        main_sizer.Add(select_label, flag=wx.EXPAND | wx.ALL, border=5)
        main_sizer.Add(location_panel, flag=wx.EXPAND | wx.ALL, border=5)
        main_sizer.Add(edit_panel, flag=wx.EXPAND|wx.ALL, border=5,
                       proportion=1)
        main_sizer.Add(control_panel, flag=wx.EXPAND|wx.ALL, border=5)



        # Location panel
        location_sizer = wx.BoxSizer(wx.HORIZONTAL)
        location_panel.SetSizer(location_sizer)
        self.path = wx.TextCtrl(location_panel,
                                value='Click to open location...',
                                style=wx.TE_READONLY)
        open_button = wx.Button(location_panel, wx.ID_OPEN)

        location_sizer.Add(self.path, proportion=1, flag=wx.GROW|wx.ALL,
                           border=5)
        location_sizer.Add(open_button, flag=wx.ALL, border=5)
        
        self.Bind(wx.EVT_BUTTON, self.click_open, open_button)

        location_panel.Show()



        # Edit panel
        edit_sizer = wx.StaticBoxSizer(wx.StaticBox(edit_panel,label=
                                        'Messages'), wx.HORIZONTAL)
        edit_panel.SetSizer(edit_sizer)
        self.message_pane = wx.TextCtrl(edit_panel, style=wx.TE_MULTILINE |
                                   wx.TE_READONLY)

        edit_sizer.Add(self.message_pane, flag=wx.EXPAND|wx.ALL, proportion=1,
                       border=5)


        # Control panel
        control_sizer = wx.BoxSizer(wx.HORIZONTAL)
        control_panel.SetSizer(control_sizer)
        self.create_button = wx.Button(control_panel, label='Cre&ate')
        self.create_button.Enable(False)
        self.edit_button = wx.Button(control_panel, label='&Edit')
        self.edit_button.Enable(False)
        self.update_button = wx.Button(control_panel, label='&Update')
        self.update_button.Enable(False)
        close_button = wx.Button(control_panel, wx.ID_CLOSE)
        about_button = wx.Button(control_panel, label='&About')

        control_sizer.Add(self.create_button, proportion=0, border=5)
        control_sizer.Add(self.edit_button, proportion=0, border=5)
        control_sizer.Add(self.update_button, proportion=0, border=5)
        control_sizer.Add(close_button, border=5)
        control_sizer.Add(about_button, flag=wx.LEFT, border=50)

        self.Bind(wx.EVT_BUTTON, self.click_create, self.create_button)
        self.Bind(wx.EVT_BUTTON, self.on_click_close, close_button)
        self.Bind(wx.EVT_BUTTON, self.click_update, self.update_button)
        self.edit_button.Bind(wx.EVT_BUTTON, self.click_edit)
        about_button.Bind(wx.EVT_BUTTON, self.click_about)

        # Resize the frame to contain all widgets
        #self.Fit()

    # Event handler for path updates in the path field
    def path_updated(self, event):
        pass
        print 'path_update'

    # Event handler for close button
    def on_click_close(self, event):
        self.Close()

    # Event handler for open button
    def click_open(self, event):
        dir_open_dialog = wx.DirDialog(self, message='Choose backup location',
                                       style=wx.DD_DEFAULT_STYLE,
                                       defaultPath="")
        resp = dir_open_dialog.ShowModal()
        print dir_open_dialog.GetPath()
        if resp == wx.ID_OK:
            print dir_open_dialog.GetPath()
            self.path.ChangeValue(dir_open_dialog.GetPath())
            if backup.is_valid_backup(self.path.GetValue()):
                self.edit_button.Enable(True)
                self.update_button.Enable(True)
            else:
                self.create_button.Enable(True)
            dir_open_dialog.Destroy()
        else:
            dir_open_dialog.Destroy()

    # Event handler for add file button
    def add_file(self, event):
        file_open_dialog = wx.FileDialog(self, style=wx.FD_OPEN,
                                         message='Choose the file to add')
        resp = file_open_dialog.ShowModal()
        if resp == wx.ID_OK:
            backup.add_path_to_data_file(self.path.GetValue(),
                                        file_open_dialog.GetPath())
            file_open_dialog.Destroy()
        else:
            file_open_dialog.Destroy()

    def click_create(self, event):
        backup.create_data_file(self.path.GetValue())
        self.edit_button.Enable(True)
        self.update_button.Enable(True)
        self.create_button.Enable(False)
        self.message_pane.AppendText('New backup location created\n')


    # Event handler for add folder button.
    def add_folder(self, event):
        pass
        dir_dialog = wx.DirDialog(self, style=wx.DD_DEFAULT_STYLE,
                                  message='Choose folder to add')
        resp = dir_dialog.ShowModal()
        if resp == wx.ID_OK:
            backup.add_path_to_data_file(self.path.GetValue(),
                                         dir_dialog.GetPath())
            dir_dialog.Destroy()
        else:
            dir_dialog.Destroy()

    # Event handler for update button
    def click_update(self, event):
        self.message_pane.AppendText('Updating backup...wait\n')
        backup.update_all(self.path.GetValue())
        self.message_pane.AppendText('Done!\n')

    # Event handler for edit button
    def click_edit(self, event):
        selected_bkp = self.path.GetValue()
        source_paths = backup.get_backup_sources(selected_bkp)
        dialog = itemdialog.ItemDialog(self,'Add item',source_paths)
        resp = dialog.ShowModal()
        if resp == wx.ID_OK:
            modified_paths = dialog.get_list()
            backup.update_config_file(selected_bkp, modified_paths)
            dialog.Destroy()
        else:
            dialog.Destroy()

    def click_about(self, event):
        abt.About()
        
# App initialiser
if __name__ == '__main__':
    app = wx.App(redirect=False)
    MainFrame()
    app.MainLoop()
